File: assignment 7 and 8 WPP/ques1.py
'''1. Write  a  Python  program  to  create  a  class  representing  a  linked  list  data  structure.  Include 
methods for displaying linked list data, inserting and deleting nodes. '''
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def display(self):
        current = self.head
        while current:
            print(current.data, end=' -> ')
            current = current.next
        print('None')

    def insert(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next: current = current.next
            current.next = new_node

    def delete(self, key):
        current = self.head
        prev = None
        
        if current and current.data == key:
            self.head = current.next
            current = None
            return
        
        while current and current.data != key:
            prev = current
            current = current.next
        
        if current is None:
            print("Key not found in the list.")
            return
        
        prev.next = current.next
        current = None

# Example Usage
ll = LinkedList()
ll.insert(10)
ll.insert(20)
ll.insert(30)
ll.display()
ll.delete(20)
ll.display()

File: assignment 4 DS/ques1.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
  int data;
  struct Node *next;
};

struct Node *head = NULL;

void insert(int data, int position)
{
  struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
  newNode->data = data;

  if (position == 1)
  {
    newNode->next = head;
    head = newNode;
    return;
  }

  struct Node *temp = head;
  for (int i = 1; i < position - 1 && temp != NULL; i++)
  {
    temp = temp->next;
  }

  if (temp == NULL)
  {
    printf("Position out of range\n");
    return;
  }

  newNode->next = temp->next;
  temp->next = newNode;
}

void delete(int position)
{
  if (head == NULL)
  {
    printf("List is empty\n");
    return;
  }

  struct Node *temp = head;

  if (position == 1)
  {
    head = head->next;
    free(temp);
    return;
  }

  for (int i = 1; i < position - 1 && temp != NULL; i++)
  {
    temp = temp->next;
  }

  if (temp == NULL || temp->next == NULL)
  {
    printf("Position out of range\n");
    return;
  }

  struct Node *nodeToDelete = temp->next;
  temp->next = nodeToDelete->next;
  free(nodeToDelete);
}

void display()
{
  struct Node *temp = head;
  while (temp != NULL)
  {
    printf("%d -> ", temp->data);
    temp = temp->next;
  }
  printf("NULL\n");
}

int main()
{
  insert(10, 1);
  insert(20, 2);
  insert(30, 3);
  display();

  delete (2);
  display();

  insert(20, 2);
  display();

  return 0;
}